### PR TITLE
Add Enhanced Broadcasting support for Linux

### PR DESCRIPTION
**Pull request by lexano-ivs**
### Description
TEB (Twitch Enhanced Broadcasting) requires the client to gather information about the host, such as which GPU(s) are available, memory sizes, OS version, etc..., and send that information to a server which then responds with a configuration set for the client to use with multitrack encoding. TEB originally supported Windows, and this PR adds support for Linux-based operating systems.

TEB on Linux supports NVIDIA and AMD GPUs, and Intel GPUs as of 2024.12.17. Intel GPUs with Linux seem to work at least with a flatpak build. For NVIDIA GPUs, NVENC is used and the configurations from the server are currently identical between Windows and Linux. For AMD GPUs, AMF is used on Windows and VAAPI is used on Linux; AMF on Linux is not currently supported for TEB. AMF on Linux might be supported in the future.

### Motivation and Context
From the client encoding perspective, Enhanced Broadcasting was originally envisioned to support Windows, Linux, and MacOS. This PR addresses the market need to expand support for Linux-based users.

### How Has This Been Tested?
The code has been tested in the TEB beta program and also by the primary developer (@lexano-ivs) on a host setup specifically to test TEB on Linux. This host includes:

- iGPU (Intel)
- NVIDIA 4060
- AMD 6650 XT and AMD 7800 XT
- Intel Arc A770
- Multi-boot system with Ubuntu 24.04, Linux Mint 22, and EndeavourOS (Arch-based)

The code was tested against multiple configurations, including single & dual dGPU configurations, swapping the primary/secondary GPU between NVIDIA and AMD, and across the 3 Linux operating systems. DEB-based builds, local builds, and flatpak builds were also tested. 

Refer to [TEB Installation Instructions](https://github.com/amazon-contributing/upstreaming-to-obs-studio/wiki/TEB-Beta-Linux-Installation) for additional context.

### Types of changes
- New feature (non-breaking change which adds functionality)

### Checklist:
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

### CODE REVIEW

File: plugins/obs-ffmpeg/obs-ffmpeg-vaapi.c
Comment (tytan652) [inline plugins/obs-ffmpeg/obs-ffmpeg-vaapi.c:1]: Changing the encoder to fit TEB is really not good at all.
Encoder properties are not API that is common between them, the change should be on TEB side not the encoder that stands out.

**Review by tytan652 — CHANGES_REQUESTED** [review]

File: UI/system-info-posix.cpp [hunk 1]
```diff
@@ -1,6 +1,623 @@
...
+
+	// Trim trailing whitespace
+	pos = s.find_last_not_of(WHITE_SPACE);
+	if (pos != string::npos)
+		s = s.substr(0, pos + 1);
+}
+
+static bool get_distribution_info(string &distro, string &version)
+{
+	ifstream file("/etc/os-release");
```
[the remark this answers is not in the document]
Comment (lexano-ivs) [reply UI/system-info-posix.cpp:65]: Thanks tytan652. Is there something else within a flatpak that can be relied upon?

File: plugins/obs-ffmpeg/obs-ffmpeg-vaapi.c
Comment (Lain-B) [reply plugins/obs-ffmpeg/obs-ffmpeg-vaapi.c:1]: I think an argument could be made that this makes the VAAPI encoder more consistent with other encoders. Our implementation of other encoders typically try to use strings rather than integers specifically to prevent issues not just with TEB, but similar situations in general where you want to set it programmatically anyway (scripts, frontends, or whatever other cases), or when you want to manually edit config files or use json directly for whatever reason. We should have had it string-based from the beginning (in my humble opinion at least).

The one issue with this change is that settings migration will have to be added for this if we want to do it this way.

So I agree with this change personally, but we need a migration path if we want to change these settings.

I'd probably advocate changing the setting's string name from `"profile"` to something like `"profile_v2"` or `"profile_str"` for migration, and then in the update function where it reads the settings, check to see if new new profile setting exists. If it does not exist, and the old setting exists, then just use that, otherwise use the new setting. At least that's what I would do. It would be simple and easy.

i.e.
```c
int profile = default_val_or_something_here_maybe;

if (!obs_data_has_user_value(settings, "profile_v2") && obs_data_has_user_value(settings, "profile")) {
    profile = obs_data_get_int(settings, "profile);
    // use the integer value
} else {
   const char *profile_str = obs_data_get_string(settings, "profile_v2");
   if (astrcmpi(profile_str, h264_main) == 0) {
       profile = AV_PROFILE_H264_MAIN;
   } else if { // etc
   }
   // parse the string value to integer
}
```
or something like that. Very rough code here, maybe somebody has a better suggestion but something like that presumably.

File: UI/system-info-posix.cpp
Comment (Lain-B) [inline UI/system-info-posix.cpp:1]: Okay, so for the all of the `UI/system-info-posix.cpp` changes, there seems to be a bit of a mix between using raw pointers and references.

```c
static void trim_ws(std::string &s)
```
```c
static bool get_distribution_info(string &distro, string &version)
```
```c
static bool get_cpu_freq(uint32_t *cpu_freq)
```
```c
static bool get_drm_card_info(uint8_t *card_idx, struct drm_card_info *dci)
```
```c
static void adjust_gpu_model(std::string *model)
```
```c
bool compare_match_strength(const drm_card_info &a, const drm_card_info &b)
```
I'm not entirely sure why pointers were used in certain cases here like this versus references. In C this would be understandable but in C++ this just adds unnecessarily unsafe code. Might want to fix these (and any other similar cases in here I haven't spotted) to use references.

If you need the ability to pass `nullptr` (which I don't quite see), that would also make sense, but I would probably try to see if that's absolutely necessary, and if so, to check to see if there's a safer way of passing that (e.g. `std::optional`).

File: UI/system-info-posix.cpp [hunk 2]
```diff
@@ -1,6 +1,623 @@
...
+	size_t last = 0;
+	first = model->find('[');
+	last = model->find_last_of(']');
+	if ((last - first) > 1) {
+		std::string adjusted_model = model->substr(first + 1, last - first - 1);
+		model->assign(adjusted_model);
+	}
+}
+
+bool compare_match_strength(const drm_card_info &a, const drm_card_info &b)
```
Comment (Lain-B) [inline UI/system-info-posix.cpp:348]: Unless this function is used in a different translation unit, might want to mark it as static.

**Review by Lain-B — CHANGES_REQUESTED** [review]
In addition to my comment on Tytan's comment about VAAPI settings migration, these are the only issues I see so far.

File: UI/system-info-posix.cpp (hunk 2 above)
Comment (Lain-B) [reply UI/system-info-posix.cpp:348]: Also recommend checking other newly added non-exported functions elsewhere where this situation might apply, but this is just the one I noticed at first glance.

File: plugins/obs-ffmpeg/obs-ffmpeg-vaapi.c
Comment (tytan652) [reply plugins/obs-ffmpeg/obs-ffmpeg-vaapi.c:1]: If it is decided/wished to keep those changes, it needs to be split in another PR with a clear motivation (and TEB is not one in my opinion).

File: UI/system-info-posix.cpp (hunk 2 above)
Comment (gxalpha) [reply UI/system-info-posix.cpp:348]: > Unless this function is used in a different translation unit, might want to mark it as static.

Or for better C++-isms, put them into an anonymous (unnamed) namespace.

File: plugins/obs-ffmpeg/obs-ffmpeg-vaapi.c
Comment (lexano-ivs) [reply plugins/obs-ffmpeg/obs-ffmpeg-vaapi.c:1]: With the latest set of commits, I reverted the VAAPI profile handling to being integer instead of string-based, and converted the incoming profile string to the corresponding integer, in the UI code. I think the VAAPI code should be using strings as well to be consistent with the other APIs. I can open another PR for this as needed.

One related issue I ran into while trying to implement a migration path is in `vaapi_device_modified()`. The function currently does not have access to the codec type, hence the [switch statement](https://github.com/obsproject/obs-studio/blob/f03280c4fd449d1591363291fd2762ffca878e6a/plugins/obs-ffmpeg/obs-ffmpeg-vaapi.c#L915) is blindly stepping through the profiles for all codecs. Eventually this might fail because the codec profile definitions are not all unique across AVC, HEVC, and AV1.

In short, there is enough work & risk here to make this a separate PR.

File: UI/system-info-posix.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (lexano-ivs) [reply UI/system-info-posix.cpp]: I've added parsing of the `/.flatpak-info` in addition to `/etc/os-release`.

File: UI/system-info-posix.cpp
Comment (lexano-ivs) [reply UI/system-info-posix.cpp:1]: Thank you. I've cleaned this up with the latest commits. It's more consistent now and using references instead of pointers.

File: UI/system-info-posix.cpp (hunk 2 above)
Comment (lexano-ivs) [reply UI/system-info-posix.cpp]: I've put almost all functions in an anonymous namespace and dropped "static" for these. It's cleaner from perspective. Thanks for the suggestion, it makes sense.

File: UI/system-info-posix.cpp [hunk 3]
```diff
@@ -1,6 +1,672 @@
...
+{
+	ifstream file;
+	std::string line;
+	const std::string systemd_file = "/etc/os-release";
+	const std::string flatpak_file = "/.flatpak-info";
+
+	distro = "";
+	version = "";
+
+	if (std::filesystem::exists(systemd_file)) {
```
Comment (tytan652) [inline UI/system-info-posix.cpp:94]: Wrong priority, Flatpak should be checked first.

Some runtime add an os-release file for compatibility, but has been proven to be an issue (e.g. 6.2 was still showing 5.15 in os-release).

Comment (farhadfarhadii) [reply UI/system-info-posix.cpp:94]:
```suggestion
	if (std::filesystem::exists(flatpak_file)) {
		/* The .flatpak-info file has a line of the form:
		 *
		 * runtime=runtime/org.kde.Platform/x86_64/6.6
		 *
		 * Parse the line into tokens to identify the name and
		 * version, which are "org.kde.Platform" and "6.6" respectively,
		 * in the example below.
		 */
		file.open(flatpak_file);
		if (file.is_open()) {
			while (getline(file, line)) {
				if (line.compare(0, 16, "runtime=runtime/") == 0) {
					size_t pos = line.find('/');
					if (pos != string::npos && line.at(pos + 1) != '\0') {
						line.erase(0, pos + 1);

						/* Split the string into tokens with a regex
						 * a regex of one or more '/' characters'.
						 */
						std::regex fp_reg("[/]+");
						vector<std::string> fp_tokens(
							sregex_token_iterator(line.begin(), line.end(), fp_reg, -1),
							sregex_token_iterator());
						if (fp_tokens.size() >= 2) {
							auto token = begin(fp_tokens);
							distro = "Flatpak " + *token;
							token = next(fp_tokens.end(), -1);
							version = *token;
						} else {
							distro = "Flatpak unknown";
							version = "0";
							blog(LOG_DEBUG,
							     "%s: Format of 'runtime' entry unrecognized in file %s",
							     __FUNCTION__, flatpak_file.c_str());
						}
						break;
					}
				}
			}
			file.close();
		}
	} else if (std::filesystem::exists(systemd_file)) {
		/* systemd-based distributions use /etc/os-release to identify
		 * the OS. For example, the Ubuntu 24.04.1 variant looks like:
		 *
		 * PRETTY_NAME="Ubuntu 24.04.1 LTS"
		 * NAME="Ubuntu"
		 * VERSION_ID="24.04"
		 * VERSION="24.04.1 LTS (Noble Numbat)"
		 * VERSION_CODENAME=noble
		 * ID=ubuntu
		 * ID_LIKE=debian
		 * HOME_URL="https://www.ubuntu.com/"
		 * SUPPORT_URL="https://help.ubuntu.com/"
		 * BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
		 * PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
		 * UBUNTU_CODENAME=noble
		 * LOGO=ubuntu-logo
		 *
		 * Parse the file looking for the NAME and VERSION_ID fields.
		 * Note that some distributions wrap the entry in '"' characters
		 * while others do not, so we need to remove those characters.
		 */
		file.open(systemd_file);
		if (file.is_open()) {
			while (getline(file, line)) {
				if (line.compare(0, 4, "NAME") == 0) {
					size_t pos = line.find('=');
					if (pos != std::string::npos && line.at(pos + 1) != '\0') {
						distro = line.substr(pos + 1);

						// Remove the '"' characters from the string, if any.
						distro.erase(std::remove(distro.begin(), distro.end(), '"'),
							     distro.end());

						trim_ws(distro);
						continue;
					}
				}

				if (line.compare(0, 10, "VERSION_ID") == 0) {
					size_t pos = line.find('=');
					if (pos != std::string::npos && line.at(pos + 1) != '\0') {
						version = line.substr(pos + 1);

						// Remove the '"' characters from the string, if any.
						version.erase(std::remove(version.begin(), version.end(), '"'),
							      version.end());

						trim_ws(version);
						continue;
					}
				}
			}
			file.close();
		}
	} else {
		blog(LOG_DEBUG, "%s: Failed to find host OS or flatpak info", __FUNCTION__);
		return false;
	}

	return true;
```

File: UI/system-info-posix.cpp [hunk 4]
```diff
@@ -1,6 +1,672 @@
...
+							blog(LOG_DEBUG,
+							     "%s: Format of 'runtime' entry unrecognized in file %s",
+							     __FUNCTION__, flatpak_file.c_str());
+						}
+						break;
+					}
+				}
+			}
+			file.close();
+		}
```
Comment (eiymba) [inline UI/system-info-posix.cpp:190]: Thanks for the PR! :heart: 

I've been using it to stream on Arch Linux with Wayland and an RTX 3060. This PR has removed the need to use Windows for enhanced broadcasting. You're awesome!

As a token of gratitude, I wanted to help by suggesting the changes asked in this section. Hopefully, you can just click a button to apply them if you wish!



```suggestion
        if (std::filesystem::exists(flatpak_file)) {
		/* The .flatpak-info file has a line of the form:
		 *
		 * runtime=runtime/org.kde.Platform/x86_64/6.6
		 *
		 * Parse the line into tokens to identify the name and
		 * version, which are "org.kde.Platform" and "6.6" respectively,
		 * in the example below.
		 */
		file.open(flatpak_file);
		if (file.is_open()) {
			while (getline(file, line)) {
				if (line.compare(0, 16, "runtime=runtime/") == 0) {
					size_t pos = line.find('/');
					if (pos != string::npos && line.at(pos + 1) != '\0') {
						line.erase(0, pos + 1);

						/* Split the string into tokens with a regex
						 * a regex of one or more '/' characters'.
						 */
						std::regex fp_reg("[/]+");
						vector<std::string> fp_tokens(
							sregex_token_iterator(line.begin(), line.end(), fp_reg, -1),
							sregex_token_iterator());
						if (fp_tokens.size() >= 2) {
							auto token = begin(fp_tokens);
							distro = "Flatpak " + *token;
							token = next(fp_tokens.end(), -1);
							version = *token;
						} else {
							distro = "Flatpak unknown";
							version = "0";
							blog(LOG_DEBUG,
							     "%s: Format of 'runtime' entry unrecognized in file %s",
							     __FUNCTION__, flatpak_file.c_str());
						}
						break;
					}
				}
			}
			file.close();
		}
	} else if (std::filesystem::exists(systemd_file)) {
		/* systemd-based distributions use /etc/os-release to identify
		 * the OS. For example, the Ubuntu 24.04.1 variant looks like:
		 *
		 * PRETTY_NAME="Ubuntu 24.04.1 LTS"
		 * NAME="Ubuntu"
		 * VERSION_ID="24.04"
		 * VERSION="24.04.1 LTS (Noble Numbat)"
		 * VERSION_CODENAME=noble
		 * ID=ubuntu
		 * ID_LIKE=debian
		 * HOME_URL="https://www.ubuntu.com/"
		 * SUPPORT_URL="https://help.ubuntu.com/"
		 * BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
		 * PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
		 * UBUNTU_CODENAME=noble
		 * LOGO=ubuntu-logo
		 *
		 * Parse the file looking for the NAME and VERSION_ID fields.
		 * Note that some distributions wrap the entry in '"' characters
		 * while others do not, so we need to remove those characters.
		 */
		file.open(systemd_file);
		if (file.is_open()) {
			while (getline(file, line)) {
				if (line.compare(0, 4, "NAME") == 0) {
					size_t pos = line.find('=');
					if (pos != std::string::npos && line.at(pos + 1) != '\0') {
						distro = line.substr(pos + 1);

						// Remove the '"' characters from the string, if any.
						distro.erase(std::remove(distro.begin(), distro.end(), '"'),
							     distro.end());

						trim_ws(distro);
						continue;
					}
				}

				if (line.compare(0, 10, "VERSION_ID") == 0) {
					size_t pos = line.find('=');
					if (pos != std::string::npos && line.at(pos + 1) != '\0') {
						version = line.substr(pos + 1);

						// Remove the '"' characters from the string, if any.
						version.erase(std::remove(version.begin(), version.end(), '"'),
							      version.end());

						trim_ws(version);
						continue;
					}
				}
			}
			file.close();
		}

File: UI/system-info-posix.cpp (hunk 3 above)
Comment (lexano-ivs) [reply UI/system-info-posix.cpp]: Thanks tytan652. I reversed the parsing order and pushed the update.

File: UI/system-info-posix.cpp (hunk 4 above)
Comment (lexano-ivs) [reply UI/system-info-posix.cpp]: Thank you for the kind words :-) I've reversed the parsing order of the files with the latest push.

**Comment by mihawk90** [comment]
Just wanted to add my feedback for future reviewers:
I've applied this to my build a couple weeks ago and just ran a 9h stream (on an AMD RX 7700XT) with 0 issues, it's working great.

The only thing that was odd was that the Twitch Inspector doesn't show any of the streams when in Bandwidth Testing mode, but it's showing them fine for a Live Broadcast. I don't think that's related to this PR though and more of an Inspector thing.

**Comment by dsaedtler** [comment]
> The only thing that was odd was that the Twitch Inspector doesn't show any of the streams when in Bandwidth Testing mode, but it's showing them fine for a Live Broadcast. I don't think that's related to this PR though and more of an Inspector thing.

Yeah in bandwidth testing mode the data doesn't actually get processed, so it doesn't log codecs etc.

File: frontend/utility/MultitrackVideoOutput.cpp [hunk 5]
```diff
@@ -224,6 +233,58 @@ static OBSEncoderAutoRelease create_video_encoder(DStr &name_buffer, size_t enco
 	dstr_printf(name_buffer, "multitrack video video encoder %zu", encoder_index);
 
 	OBSDataAutoRelease encoder_settings = obs_data_create_from_json(encoder_config.settings.dump().c_str());
+
+	/* VAAPI-based encoders unfortunately use an integer for "profile".
+	 * Until a string-based "profile" can be used with VAAPI, find the
+	 * corresponding integer value and update the settings with an
+	 * integer-based "profile".
+	 */
```
Comment (RytoEX) [inline frontend/utility/MultitrackVideoOutput.cpp:241]: Lines can be wrapped at 120 columns.

File: frontend/utility/MultitrackVideoOutput.cpp [hunk 6]
```diff
@@ -224,6 +233,58 @@ static OBSEncoderAutoRelease create_video_encoder(DStr &name_buffer, size_t enco
...
+	if (strstr(encoder_type, "vaapi")) {
+		// Move the "profile" string to "profile_str"
```
Comment (RytoEX) [inline frontend/utility/MultitrackVideoOutput.cpp:243]: For comments using complete sentences, add final punctuation at the end.

File: libobs-opengl/gl-subsystem.c [hunk 7]
```diff
@@ -214,6 +214,53 @@ const char *device_preprocessor_name(void)
...
+}
+
+const char *gpu_get_renderer(void)
+{
+	return ((const char *)glGetString(GL_RENDERER));
+}
+
+/* Get the amount of dedicated GDDR memory, aka VRAM, in
+ * units of kilobytes.
+ */
```
Comment (RytoEX) [inline libobs-opengl/gl-subsystem.c:229]: Lines can be wrapped at 120 columns.

File: libobs-opengl/gl-subsystem.c [hunk 8]
```diff
@@ -214,6 +214,53 @@ const char *device_preprocessor_name(void)
...
+
+	/* GLint is signed, however it makes little sense to have
+	 * "negative" amounts of GPU memory. Check on this, clamp to
+	 * 0 if so, and cast to an unsigned value explicitly.
+	 */
+	if (dmem < 0)
+		dmem = 0;
+	if (total_mem < 0)
+		total_mem = 0;
+	return (uint64_t)total_mem - dmem;
```
Comment (RytoEX) [inline libobs-opengl/gl-subsystem.c:261]: Prefer a blank line after an un-bracketed `if`, or enclose all `if`s in braces.

**Review by RytoEX — CHANGES_REQUESTED** [review]

File: UI/system-info-posix.cpp (hunk 3 above)
Comment (RytoEX) [reply UI/system-info-posix.cpp]: @tytan652 has this been addressed?

File: UI/system-info-posix.cpp (hunk 4 above)
Comment (RytoEX) [reply UI/system-info-posix.cpp]: Looks like this has been addressed.

File: plugins/obs-ffmpeg/obs-ffmpeg-vaapi.c
Comment (RytoEX) [reply plugins/obs-ffmpeg/obs-ffmpeg-vaapi.c:1]: The VAAPI changes have been removed, so it seems this has been addressed.

File: UI/system-info-posix.cpp
Comment (RytoEX) [reply UI/system-info-posix.cpp:1]: @Lain-B Looks like these have been addressed?

File: UI/system-info-posix.cpp (hunk 3 above)
Comment (tytan652) [reply UI/system-info-posix.cpp]: Yes

File: UI/system-info-posix.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (RytoEX) [reply UI/system-info-posix.cpp]: It looks like this is addressed now. @tytan652 ?

[the remark this answers is not in the document]
Comment (tytan652) [reply UI/system-info-posix.cpp]: Yes

File: plugins/obs-ffmpeg/obs-ffmpeg-vaapi.c [hunk 9]
```diff
@@ -229,7 +229,7 @@ static const rc_mode_t *get_rc_mode(const char *name)
 
 	const rc_mode_t *rc_mode = RC_MODES;
 
-	while (!!rc_mode->name && strcmp(rc_mode->name, name) != 0)
+	while (!!rc_mode->name && astrcmpi(rc_mode->name, name) != 0)
```
[the remark this answers is not in the document]
Comment (lexano-ivs) [reply plugins/obs-ffmpeg/obs-ffmpeg-vaapi.c:232]: `get_rc_mode()` is stepping through the `RC_MODES[]` array that defines which rate control modes are valid. The string in the array is uppercase (e.g. "CBR") and if a request comes in with "cbr", the function will return the `{0}` entry, which causes an immediate crash upon de-reference. Making the comparison string insensitive solves the problem.

The cases where this might happen are:

1. Someone manually entering the rc_mode using the advanced edit box in the GUI; e.g. they type "cbr" or "Cbr" instead of "CBR"
2. Enhanced Broadcasting: The GetClientConfiguration response contains rc_mode in the JSON, and we've made sure (after this bug) to always send uppercase, but in case it gets changed in the future, this will crash unless the comparison is case insensitive.

File: libobs/graphics/graphics-imports.c [hunk 10]
```diff
@@ -42,6 +42,10 @@ bool load_graphics_imports(struct gs_exports *exports, void *module, const char
 	bool success = true;
 
 	GRAPHICS_IMPORT(device_get_name);
+	GRAPHICS_IMPORT_OPTIONAL(gpu_get_smem);
+	GRAPHICS_IMPORT_OPTIONAL(gpu_get_dmem);
+	GRAPHICS_IMPORT_OPTIONAL(gpu_get_driver_version);
+	GRAPHICS_IMPORT_OPTIONAL(gpu_get_renderer);
```
[the remark this answers is not in the document]
Comment (lexano-ivs) [reply libobs/graphics/graphics-imports.c]: Fixed with latest push.

File: libobs/graphics/graphics.c [hunk 11]
```diff
@@ -297,6 +297,26 @@ const char *gs_get_device_name(void)
...
+const char *gs_get_driver_version(void)
+{
+	return gs_valid("gs_get_driver_version") ? thread_graphics->exports.gpu_get_driver_version() : NULL;
+}
+
+const char *gs_get_renderer(void)
+{
+	return gs_valid("gs_get_renderer") ? thread_graphics->exports.gpu_get_renderer() : NULL;
+}
+
```
[the remark this answers is not in the document]
Comment (lexano-ivs) [reply libobs/graphics/graphics.c]: Fixed with latest push.

File: UI/window-basic-settings.cpp [hunk 12]
```diff
@@ -5616,7 +5616,10 @@ void OBSBasicSettings::UpdateMultitrackVideo()
 			ui->enableMultitrackVideo->setChecked(false);
 	}
 
-#ifndef _WIN32
+	// Enhanced Broadcasting works on Windows and Linux. For other
+	// OS variants, only enable the GUI controls if developer mode
+	// was invoked.
+#if (!defined(__linux__) && !defined(_WIN32))
```
[the remark this answers is not in the document]
Comment (lexano-ivs) [reply UI/window-basic-settings.cpp]: Yup, good catch. I've rebased to `master` as of about 1 hour ago and fixed the conflict.

File: frontend/settings/OBSBasicSettings.cpp [hunk 13]
```diff
@@ -5588,7 +5588,10 @@ void OBSBasicSettings::UpdateMultitrackVideo()
 			ui->enableMultitrackVideo->setChecked(false);
 	}
 
-#ifndef _WIN32
+	// Enhanced Broadcasting works on Windows and Linux. For other
+	// OS variants, only enable the GUI controls if developer mode
+	// was invoked.
+#if (!defined(__linux__) && !defined(_WIN32))
```
[the remark this answers is not in the document]
Comment (lexano-ivs) [reply frontend/settings/OBSBasicSettings.cpp]: Fixed with latest push.

File: libobs-opengl/gl-subsystem.c [hunk 14]
```diff
@@ -214,6 +214,53 @@ const char *device_preprocessor_name(void)
...
+	GLint dmem = 0;
+	glGetIntegerv(GL_GPU_MEMORY_INFO_DEDICATED_VIDMEM_NVX, &dmem);
+
+	/* GLint is signed, however it makes little sense to have
+	 * "negative" amounts of GPU memory. Check on this, clamp to
+	 * 0 if so, and cast to an unsigned value explicitly.
+	 */
+	if (dmem < 0)
+		dmem = 0;
+	return (uint64_t)dmem;
```
[the remark this answers is not in the document]
Comment (lexano-ivs) [reply libobs-opengl/gl-subsystem.c:240]: Fixed with latest push.

File: frontend/utility/MultitrackVideoOutput.cpp (hunk 5 above)
Comment (lexano-ivs) [reply frontend/utility/MultitrackVideoOutput.cpp]: Fixed with latest push.

File: frontend/utility/MultitrackVideoOutput.cpp (hunk 6 above)
Comment (lexano-ivs) [reply frontend/utility/MultitrackVideoOutput.cpp]: Fixed with latest push.

File: libobs-opengl/gl-subsystem.c (hunk 7 above)
Comment (lexano-ivs) [reply libobs-opengl/gl-subsystem.c]: Fixed with latest push.

File: libobs-opengl/gl-subsystem.c (hunk 8 above)
Comment (lexano-ivs) [reply libobs-opengl/gl-subsystem.c:260]: Fixed with latest push.

File: plugins/obs-ffmpeg/obs-ffmpeg-vaapi.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (tytan652) [reply plugins/obs-ffmpeg/obs-ffmpeg-vaapi.c:232]: I thought about it, and unfortunately I don't think it's up to the encoder/plugin to deal with UI/frontend issues.
1. The frontend itself should prevent that.
2. The issue is on the remote if uppercase is expected by the JSON schema, if you want to workaround that it's still in the frontend.

[the remark this answers is not in the document]
Comment (RytoEX) [reply plugins/obs-ffmpeg/obs-ffmpeg-vaapi.c:232]: For what it's worth, `obs-ffmpeg`, `obs-nvenc`, and `obs-qsv11` do case-insensitive comparisons on rate control modes. Conversely, to the concerns above, the frontend itself shouldn't have to know what each of the plugins expects for string case.

[the remark this answers is not in the document]
Comment (Lain-B) [reply plugins/obs-ffmpeg/obs-ffmpeg-vaapi.c:232]: Case-insensitive for these values is typically fine because people might edit the config files themselves. Not sure about the scheme and all that though. I wasn't a part of that originally

**Review by Lain-B — APPROVED** [review]

**Review by RytoEX — APPROVED** [review]
Seems fine.
